Terraform rpc error + panic runtime error when trying to create QnAMaker with azurerm_cognitive_account resource

Hi there, I'm running into some issues when using Terraform to deploy infrastructure on Azure. Ich was trying to create a QnAMaker with Terraform. Please find all required information below.

Thanks a lot for the help in advance.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.13.2
+ provider registry.terraform.io/hashicorp/azuread v1.0.0
+ provider registry.terraform.io/hashicorp/azurerm v2.26.0
```

### Affected Resource(s)

* `azurerm_cognitive_account`

### Terraform Configuration Files

```hcl
resource "azurerm_app_service_plan" "qnamaker" {
  name                = format("%s-%s-qna", var.project, terraform.workspace)
  location            = azurerm_resource_group.project.location
  resource_group_name = azurerm_resource_group.project.name
  kind = "app"

  sku {
    tier = "Standard"
    size = "S1"
  }
}

resource "azurerm_app_service" "qnamaker" {
  name                = format("%s-%s-qna", var.project, terraform.workspace)
  location            = azurerm_resource_group.project.location
  resource_group_name = azurerm_resource_group.project.name
  app_service_plan_id = azurerm_app_service_plan.qnamaker.id
  client_affinity_enabled = true

  site_config {
    dotnet_framework_version = "v4.0"
    php_version = "5.6"
    remote_debugging_enabled = false
    use_32_bit_worker_process = true
    websockets_enabled = false
    always_on = false
    managed_pipeline_mode = "Integrated"
    default_documents = [
                    "Default.htm",
                    "Default.html",
                    "Default.asp",
                    "index.htm",
                    "index.html",
                    "iisstart.htm",
                    "default.aspx",
                    "index.php",
                    "hostingstart.html"
    ]
    cors {
      allowed_origins = [
        "*"
      ]
    }
  }

  tags = {
    format("hidden-related:%s", azurerm_app_service_plan.qnamaker.id): "empty"
  }
}

resource "azurerm_search_service" "qnamaker" {
  name                = format("%s-%s-qna", var.project, terraform.workspace)
  resource_group_name = azurerm_resource_group.project.name
  location            = azurerm_resource_group.project.location
  sku                 = "standard"
}

resource "azurerm_cognitive_account" "qnamaker" {
  name                = format("%s-%s-qna", var.project, terraform.workspace)
  location            = azurerm_resource_group.project.location
  resource_group_name = azurerm_resource_group.project.name
  kind                = "QnAMaker"
  sku_name = "S0"
  qna_runtime_endpoint = format("https://%s.azurewebsites.net", azurerm_app_service.qnamaker.name)
}
```

### Debug Output

https://gist.github.com/kevroes/3967225fb7a3324d637cd0c0741d4d15

### Panic Output

https://gist.github.com/kevroes/b7c084636eac7f1d34d0a9bd7cf59a29

### Expected Behavior

The Cognitive Service Account QnAMaker should have been created, using the App Service as QnAMaker Runtime. No Error should occur.

### Actual Behavior

While creating the Cognitive Service Account (kind: "QnAMaker") Terraform crashed, with a rpc error and panic runtime error. The resource hasn't been created.

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Running on local machine in Windows Powershell.

Relevant section from the crash log:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x8 pc=0x352c67b]

goroutine 116 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/cognitive.resourceArmCognitiveAccountCreate(0xc0009dd1f0, 0x40a68e0, 0xc000176280, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/cognitive/cognitive_account_resource.go:165 +0x5cb
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc0004a3e60, 0xc000796f50, 0xc000db0780, 0x40a68e0, 0xc000176280, 0x40cdc01, 0xc00050b208, 0xc00081db00)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:310 +0x36c
```

--- a/azurerm/internal/services/cognitive/cognitive_account_resource.go
+++ b/azurerm/internal/services/cognitive/cognitive_account_resource.go
@@ -208,12 +208,12 @@ func resourceArmCognitiveAccountUpdate(d *schema.ResourceData, meta interface{})
 		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
 	}
 
-	if d.Get("kind").(string) == "QnAMaker" && *props.Properties.APIProperties.QnaRuntimeEndpoint == "" {
-		return fmt.Errorf("the QnAMaker runtime endpoint `qna_runtime_endpoint` is required when kind is set to `QnAMaker`")
-	}
-
-	if v, ok := d.GetOk("qna_runtime_endpoint"); ok {
-		props.Properties.APIProperties.QnaRuntimeEndpoint = utils.String(v.(string))
+	if kind := d.Get("kind"); kind == "QnAMaker" {
+		if v, ok := d.GetOk("qna_runtime_endpoint"); ok && v != "" {
+			props.Properties.APIProperties.QnaRuntimeEndpoint = utils.String(v.(string))
+		} else {
+			return fmt.Errorf("the QnAMaker runtime endpoint `qna_runtime_endpoint` is required when kind is set to `QnAMaker`")
+		}
 	}
 
 	if _, err = client.Update(ctx, id.ResourceGroup, id.Name, props); err != nil {

--- a/azurerm/internal/services/cognitive/cognitive_account_resource.go
+++ b/azurerm/internal/services/cognitive/cognitive_account_resource.go
@@ -138,7 +138,7 @@ func resourceArmCognitiveAccountCreate(d *schema.ResourceData, meta interface{})
 		existing, err := client.GetProperties(ctx, resourceGroup, name)
 		if err != nil {
 			if !utils.ResponseWasNotFound(existing.Response) {
-				return fmt.Errorf("Error checking for presence of existing Cognitive Account %q (Resource Group %q): %s", name, resourceGroup, err)
+				return fmt.Errorf("checking for presence of existing Cognitive Account %q (Resource Group %q): %s", name, resourceGroup, err)
 			}
 		}
 
@@ -149,7 +149,7 @@ func resourceArmCognitiveAccountCreate(d *schema.ResourceData, meta interface{})
 
 	sku, err := expandAccountSkuName(d.Get("sku_name").(string))
 	if err != nil {
-		return fmt.Errorf("error expanding sku_name for Cognitive Account %s (Resource Group %q): %v", name, resourceGroup, err)
+		return fmt.Errorf("expanding sku_name for Cognitive Account %s (Resource Group %q): %v", name, resourceGroup, err)
 	}
 
 	props := cognitiveservices.Account{
@@ -162,21 +162,21 @@ func resourceArmCognitiveAccountCreate(d *schema.ResourceData, meta interface{})
 		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
 	}
 
-	if kind == "QnAMaker" && *props.Properties.APIProperties.QnaRuntimeEndpoint == "" {
-		return fmt.Errorf("the QnAMaker runtime endpoint `qna_runtime_endpoint` is required when kind is set to `QnAMaker`")
-	}
-
-	if v, ok := d.GetOk("qna_runtime_endpoint"); ok {
-		props.Properties.APIProperties.QnaRuntimeEndpoint = utils.String(v.(string))
+	if kind == "QnAMaker" {
+		if v, ok := d.GetOk("qna_runtime_endpoint"); ok && v != "" {
+			props.Properties.APIProperties.QnaRuntimeEndpoint = utils.String(v.(string))
+		} else {
+			return fmt.Errorf("the QnAMaker runtime endpoint `qna_runtime_endpoint` is required when kind is set to `QnAMaker`")
+		}
 	}
 
 	if _, err := client.Create(ctx, resourceGroup, name, props); err != nil {
-		return fmt.Errorf("Error creating Cognitive Services Account %q (Resource Group %q): %+v", name, resourceGroup, err)
+		return fmt.Errorf("creating Cognitive Services Account %q (Resource Group %q): %+v", name, resourceGroup, err)
 	}
 
 	read, err := client.GetProperties(ctx, resourceGroup, name)
 	if err != nil {
-		return fmt.Errorf("Error retrieving Cognitive Services Account %q (Resource Group %q): %+v", name, resourceGroup, err)
+		return fmt.Errorf("retrieving Cognitive Services Account %q (Resource Group %q): %+v", name, resourceGroup, err)
 	}
 
 	d.SetId(*read.ID)

--- a/azurerm/internal/services/cognitive/cognitive_account_resource.go
+++ b/azurerm/internal/services/cognitive/cognitive_account_resource.go
@@ -57,6 +57,7 @@ func resourceArmCognitiveAccount() *schema.Resource {
 				ForceNew: true,
 				ValidateFunc: validation.StringInSlice([]string{
 					"Academic",
+					"AnomalyDetector",
 					"Bing.Autosuggest",
 					"Bing.Autosuggest.v7",
 					"Bing.CustomSearch",

--- a/azurerm/internal/services/cognitive/tests/cognitive_account_resource_test.go
+++ b/azurerm/internal/services/cognitive/tests/cognitive_account_resource_test.go
@@ -3,6 +3,7 @@ package tests
 import (
 	"fmt"
 	"net/http"
+	"regexp"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
@@ -174,6 +175,22 @@ func TestAccAzureRMCognitiveAccount_qnaRuntimeEndpoint(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMCognitiveAccount_qnaRuntimeEndpointUnspecified(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_cognitive_account", "test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: testCheckAzureRMAppCognitiveAccountDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccAzureRMCognitiveAccount_qnaRuntimeEndpointUnspecified(data),
+				ExpectError: regexp.MustCompile("the QnAMaker runtime endpoint `qna_runtime_endpoint` is required when kind is set to `QnAMaker`"),
+			},
+		},
+	})
+}
+
 func TestAccAzureRMCognitiveAccount_cognitiveServices(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_cognitive_account", "test")
 
@@ -265,8 +282,7 @@ resource "azurerm_cognitive_account" "test" {
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
   kind                = "Face"
-
-  sku_name = "S0"
+  sku_name            = "S0"
 }
 `, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
 }
@@ -287,8 +303,7 @@ resource "azurerm_cognitive_account" "test" {
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
   kind                = "SpeechServices"
-
-  sku_name = "S0"
+  sku_name            = "S0"
 }
 `, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
 }
@@ -303,8 +318,7 @@ resource "azurerm_cognitive_account" "import" {
   location            = azurerm_cognitive_account.test.location
   resource_group_name = azurerm_cognitive_account.test.resource_group_name
   kind                = azurerm_cognitive_account.test.kind
-
-  sku_name = "S0"
+  sku_name            = "S0"
 }
 `, template)
 }
@@ -325,8 +339,7 @@ resource "azurerm_cognitive_account" "test" {
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
   kind                = "Face"
-
-  sku_name = "S0"
+  sku_name            = "S0"
 
   tags = {
     Acceptance = "Test"
@@ -347,16 +360,35 @@ resource "azurerm_resource_group" "test" {
 }
 
 resource "azurerm_cognitive_account" "test" {
+  name                 = "acctestcogacc-%d"
+  location             = azurerm_resource_group.test.location
+  resource_group_name  = azurerm_resource_group.test.name
+  kind                 = "QnAMaker"
+  qna_runtime_endpoint = "%s"
+  sku_name             = "S0"
+}
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, url)
+}
+
+func testAccAzureRMCognitiveAccount_qnaRuntimeEndpointUnspecified(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-cognitive-%d"
+  location = "%s"
+}
+
+resource "azurerm_cognitive_account" "test" {
   name                = "acctestcogacc-%d"
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
-
-  kind                 = "QnAMaker"
-  qna_runtime_endpoint = "%s"
-
-  sku_name = "S0"
+  kind                = "QnAMaker"
+  sku_name            = "S0"
 }
-`, data.RandomInteger, data.Locations.Ternary, data.RandomInteger, url)
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
 }
 
 func testAccAzureRMCognitiveAccount_cognitiveServices(data acceptance.TestData) string {

--- a/website/docs/r/cognitive_account.html.markdown
+++ b/website/docs/r/cognitive_account.html.markdown
@@ -42,7 +42,7 @@ The following arguments are supported:
 
 * `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
 
-* `kind` - (Required) Specifies the type of Cognitive Service Account that should be created. Possible values are `Academic`, `Bing.Autosuggest`, `Bing.Autosuggest.v7`, `Bing.CustomSearch`, `Bing.Search`, `Bing.Search.v7`, `Bing.Speech`, `Bing.SpellCheck`, `Bing.SpellCheck.v7`, `CognitiveServices`, `ComputerVision`, `ContentModerator`, `CustomSpeech`, `CustomVision.Prediction`, `CustomVision.Training`, `Emotion`, `Face`,`FormRecognizer`, `ImmersiveReader`, `LUIS`, `LUIS.Authoring`, `QnAMaker`, `Recommendations`, `SpeakerRecognition`, `Speech`, `SpeechServices`, `SpeechTranslation`, `TextAnalytics`, `TextTranslation` and `WebLM`. Changing this forces a new resource to be created.
+* `kind` - (Required) Specifies the type of Cognitive Service Account that should be created. Possible values are `Academic`, `AnomalyDetector`, `Bing.Autosuggest`, `Bing.Autosuggest.v7`, `Bing.CustomSearch`, `Bing.Search`, `Bing.Search.v7`, `Bing.Speech`, `Bing.SpellCheck`, `Bing.SpellCheck.v7`, `CognitiveServices`, `ComputerVision`, `ContentModerator`, `CustomSpeech`, `CustomVision.Prediction`, `CustomVision.Training`, `Emotion`, `Face`,`FormRecognizer`, `ImmersiveReader`, `LUIS`, `LUIS.Authoring`, `QnAMaker`, `Recommendations`, `SpeakerRecognition`, `Speech`, `SpeechServices`, `SpeechTranslation`, `TextAnalytics`, `TextTranslation` and `WebLM`. Changing this forces a new resource to be created.
 
 * `sku_name` - (Required) Specifies the SKU Name for this Cognitive Service Account. Possible values are `F0`, `F1`, `S0`, `S1`, `S2`, `S3`, `S4`, `S5`, `S6`, `P0`, `P1`, and `P2`.
 

